db/channel_cache.go

```diff
@@ -17,11 +17,78 @@ var (
 
 const NoSeq = uint64(0x7FFFFFFFFFFFFFFF)
 
+// Minimizes need to perform GSI/View queries to respond to the changes feed by keeping a per-channel
+// cache of the most recent changes in the channel.  Changes might be received over DCP won't necessarily
+// be in sequence order, but the changes are buffered and re-ordered before inserting into the cache, and
+// so the cache is always kept in ascending order by sequence.  (this is the global sync sequence, not the
+// per-vbucket sequence).
+//
+// Uniqueness guarantee: a given document should only have _one_ entry in the cache, which represents the
+// most recent revision (the revision with the highest sequence number).
+//
+// Completeness guarantee: the cache is guaranteed to have _every_ change in a channel from the validFrom sequence
+// up to the current known latest change in that channel.  Eg, there are no gaps/holes.
+//
+// The validFrom state variable tracks the oldest sequence for which the cache is valid, and tries to minimize
+// the number of GSI/View backfill queries by remembering the starting point of previous backfill queries.  It
+// tracks the starting point after which the cache can guarantee completeness.  As the new entries are added and
+// capacity is exceeded, the validFrom sequence will continually increase.
```

Comment:
The 'tries to' phrasing sounds wishy-washy, and makes this sound excessively complex.  I think this could be simplified to:
```
// The validFrom state variable tracks the oldest sequence for which the cache is complete.  This may be earlier 
// than the oldest entry in the cache.  Used to determine whether a GSI/View query is required for a given 
// getChanges request.
```

Reply:
Fixed